### ISSUE TITLE
All products appear in each category when category filter disabled (2.4 only)

### ISSUE THREAD

**Issue by OvalMedia** [Smile-SA/elasticsuite#1925]
Once catalog/search/engine is set to "elasticsuite" all products are visible in every category.

### Preconditions
Magento Version : 2.4 CE
ElasticSuite Version : 2.10
Elasticsearch: 7.8.1
PHP: 7.3.20

Multi-Store setup (3 websites)

Environment : developer

### Steps to reproduce
1. Create a couple of categories
2. Create a product and link it to one or more categories (not all categories though)
3. Set Catalog -> Catalog -> Catalog Search -> Search Engine to "ElasticSuite"
4. Set "Display Category Filter" (Config -> Catalog -> Catalog -> Layered Navigation) to "No"
5. Reindex everything
6. Flush cache
7. Visit categories in frontend

### Expected result
1. Categories should display the products that are linked to them

### Actual result
1. Each category shows all products of the catalog

Once I disable the smile modules and revert back to "Elasticsearch 7" search engine the store works as expected.

**EDIT:** updated description to include the crucial part (step 4).

**Comment by romainruaud** [comment]
Duplicate of #1903 #1912 and many others, scheduled to be fixed asap in next minor releases.

Regards

**Comment by romainruaud** [comment]
Please upgrade to 2.10.1 which contains the fix

**Comment by OvalMedia** [comment]
No, the problem persists.

1. upgraded to 2.10.1
2. killed all ES indexes: curl -XDELETE localhost:9200/*
3. bin/magento cache:flush
4. bin/magento indexer:reindex

Still all products are available in all categories.

**Comment by romainruaud** [comment]
Well, I might have read your issue a bit too quick because of all the other issues occuring with paginations on 2.10.0 and it looks like I misunderstood what you were explaining here, being that even products not attached to the current category are appearing into it. 

It looks like a different issue here. However, I'm actually unable to reproduce it on my Magento 2.4 with ElasticSuite 2.10.

So could you provide more details about your Magento instance, especially if you are using any third-party extension.

Also please provide an extract of the queries that are being sent to ES when you are browsing categories. You can get it from the system.log file if you activate the Debug mode of Elasticsuite in **Stores > Configuration > Elasticsuite > Debug mode**

Regards

**Comment by OvalMedia** [comment]
Yes, I use third-part extensions. But I ran into this bug last week and pulled my hair out. So I decided to rebuild this store step by step starting with the elasticsuite. I can rule out any third party software as a cause.

This is my system.log after accessing a category:

`[2020-08-24 14:43:06] main.INFO: Request Success: {"method":"HEAD","uri":"http://localhost/prefix_my_storeview_catalog_product","port":9200,"headers":{"Host":["localhost"],"Content-Type":["application/json"],"Accept":["application/json"],"User-Agent":["elasticsearch-php/7.7.0 (Linux 4.9.0-8-amd64; PHP 7.3.20-1+0~20200710.65+debian9~1.gbpc9cbeb)"]},"HTTP code":200,"duration":0.005201} []
[2020-08-24 14:43:06] main.INFO: curl -XHEAD 'http://localhost/prefix_my_storeview_catalog_product?pretty=true' [] []
[2020-08-24 14:43:06] main.INFO: Request Success: {"method":"GET","uri":"http://localhost/","port":9200,"headers":{"Host":["localhost"],"Content-Type":["application/json"],"Accept":["application/json"],"User-Agent":["elasticsearch-php/7.7.0 (Linux 4.9.0-8-amd64; PHP 7.3.20-1+0~20200710.65+debian9~1.gbpc9cbeb)"]},"HTTP code":200,"duration":0.000196} []
[2020-08-24 14:43:06] main.INFO: curl -XGET 'http://localhost/?pretty=true' [] []
[2020-08-24 14:43:06] main.INFO: Request Success: {"method":"POST","uri":"http://localhost/prefix_my_storeview_catalog_product/_search","port":9200,"headers":{"Host":["localhost"],"Content-Type":["application/json"],"Accept":["application/json"],"User-Agent":["elasticsearch-php/7.7.0 (Linux 4.9.0-8-amd64; PHP 7.3.20-1+0~20200710.65+debian9~1.gbpc9cbeb)"]},"HTTP code":200,"duration":0.00191} []
[2020-08-24 14:43:06] main.INFO: curl -XPOST 'http://localhost/prefix_my_storeview_catalog_product/_search?pretty=true' -d '{"size":0,"query":{"constant_score":{"filter":{"bool":{"must":[{"term":{"stock.is_in_stock":{"value":true,"boost":1}}},{"terms":{"visibility":[2,4],"boost":1}},{"bool":{"must_not":[{"nested":{"path":"category","score_mode":"none","query":{"bool":{"must":[{"term":{"category.category_id":{"value":7,"boost":1}}},{"term":{"category.is_blacklisted":{"value":true,"boost":1}}}],"must_not":[],"should":[],"boost":1}},"boost":1}}],"boost":1}},{"terms":{"visibility":[2,4],"boost":1}}],"must_not":[],"should":[],"boost":1}},"boost":1}},"aggregations":{"attribute_set_id":{"terms":{"field":"attribute_set_id","size":100000,"order":{"_count":"desc"}}},"indexed_attributes":{"terms":{"field":"indexed_attributes","size":100000,"order":{"_count":"desc"}}}},"track_total_hits":true}' [] []
[2020-08-24 14:43:06] main.INFO: Request Success: {"method":"POST","uri":"http://localhost/prefix_my_storeview_catalog_product/_search","port":9200,"headers":{"Host":["localhost"],"Content-Type":["application/json"],"Accept":["application/json"],"User-Agent":["elasticsearch-php/7.7.0 (Linux 4.9.0-8-amd64; PHP 7.3.20-1+0~20200710.65+debian9~1.gbpc9cbeb)"]},"HTTP code":200,"duration":0.003111} []
[2020-08-24 14:43:06] main.INFO: curl -XPOST 'http://localhost/prefix_my_storeview_catalog_product/_search?pretty=true' -d '{"size":12,"sort":[{"category.position":{"order":"asc","missing":"_last","unmapped_type":"keyword","nested":{"path":"category","filter":{"terms":{"category.category_id":["7"],"boost":1}}},"mode":"min"}},{"_score":{"order":"desc"}},{"entity_id":{"order":"desc","missing":"_first","unmapped_type":"keyword"}}],"from":0,"query":{"constant_score":{"filter":{"bool":{"must":[{"term":{"stock.is_in_stock":{"value":true,"boost":1}}},{"terms":{"visibility":[2,4],"boost":1}},{"bool":{"must_not":[{"nested":{"path":"category","score_mode":"none","query":{"bool":{"must":[{"term":{"category.category_id":{"value":7,"boost":1}}},{"term":{"category.is_blacklisted":{"value":true,"boost":1}}}],"must_not":[],"should":[],"boost":1}},"boost":1}}],"boost":1}},{"terms":{"visibility":[2,4],"boost":1}}],"must_not":[],"should":[],"boost":1}},"boost":1}},"aggregations":{"attribute_set_id":{"terms":{"field":"attribute_set_id","size":100000,"order":{"_count":"desc"}}},"categories":{"filters":{"filters":{"28":{"nested":{"path":"category","score_mode":"none","query":{"bool":{"must":[{"bool":{"must_not":[{"term":{"category.is_virtual":{"value":true,"boost":1}}}],"boost":1}},{"terms":{"category.category_id":["28"],"boost":1}}],"must_not":[],"should":[],"boost":1}},"boost":1}},"29":{"nested":{"path":"category","score_mode":"none","query":{"bool":{"must":[{"bool":{"must_not":[{"term":{"category.is_virtual":{"value":true,"boost":1}}}],"boost":1}},{"terms":{"category.category_id":["29"],"boost":1}}],"must_not":[],"should":[],"boost":1}},"boost":1}},"30":{"nested":{"path":"category","score_mode":"none","query":{"bool":{"must":[{"bool":{"must_not":[{"term":{"category.is_virtual":{"value":true,"boost":1}}}],"boost":1}},{"terms":{"category.category_id":["30"],"boost":1}}],"must_not":[],"should":[],"boost":1}},"boost":1}},"31":{"nested":{"path":"category","score_mode":"none","query":{"bool":{"must":[{"bool":{"must_not":[{"term":{"category.is_virtual":{"value":true,"boost":1}}}],"boost":1}},{"terms":{"category.category_id":["31"],"boost":1}}],"must_not":[],"should":[],"boost":1}},"boost":1}}}}},"price.price":{"nested":{"path":"price"},"aggregations":{"price.price":{"filter":{"terms":{"price.customer_group_id":[0],"boost":1}},"aggregations":{"price.price":{"histogram":{"field":"price.price","interval":1,"min_doc_count":1}}}}}}},"track_total_hits":true}' [] []
`

**Comment by OvalMedia** [comment]
I have to paddle back a little for now. A separate test system does not show the same symptoms. Will keep you updated.

**Comment by bmelman** [comment]
Had the same issue on categories that were not anchor: had to disable elasticsuite (since it removes attribute is anchor option, without enabling them), set all categories is_anchor=1, reenable elasticsuite, reindex AND save all impacted categories in admin.

**Comment by OvalMedia** [comment]
@bmelman Thanks, but that was not it. In fact I created all categories manually AFTER installing the elasticsearch suite. I disabled all smile* modules and checked the anchor status. ALL categories have this option enabled. But you made me think. What if the root categories are anchors as well? This would totally match the behaviour. Indeed, all roto categories where anchors.
So I disabled the anchor status for the root categories only and tried again. Unfortunately it did not change a thing. I also checked AGAIN to see if the smile modules secretely made the root categories anchors again. But that was not the case.

**Comment by choleaoum** [comment]
What theme are you running? After I applied the update, it still happened when I made any text search queries. Turns out it was an issue with my theme.

**Comment by OvalMedia** [comment]
Custom theme from scratch. I applied it to the test system but it is not the cause for the error.

**Comment by OvalMedia** [comment]
After rebuilding from scratch while testing and verifying each step of the way I finally found the bug (?).
I disabled the category filter (general -> catalog -> layered navigation -> Display Category Filter) and all products show inside each category.

**Comment by choleaoum** [comment]
It's most likely your third party layered navigation module, that was also an issue for me.

**Comment by romainruaud** [comment]
Interesting, I was not aware of this "Display Category Filter" feature.

Is it something that Magento introduced in 2.4 or rather something that is coming from a third party module ?

**Comment by OvalMedia** [comment]
No third party module involved. Must be a new feature/option in 2.4.

**Comment by OvalMedia** [comment]
https://github.com/magento/magento2/pull/27165

**Comment by choleaoum** [comment]
Very strange, I'm also running 2.4 but I don't have that feature.

**Comment by OvalMedia** [comment]
I switched to Luma and the result is the same. Read the pull request, I figure that will give some insight as to why this effects your module. It seems this feature is removing the filters with a plugin "afterGetFilters". So they let magento to the hard work of assembling the filters only to remove them after that. I can only assume your module is somewhat relying on them still being there.

**Comment by romainruaud** [comment]
Okay, we were not aware of this feature.

We'll proceed with this, thank you for reporting it

**Comment by nkajic** [comment]
Hi all, just a one hint to share regarding this problem. 

This problem is not only triggered by "Display Category Filter" but also on last level categories (with no childrens) where filter is not shown because there is nothing to filter. In that case there will also be shown all products. M 2.4.0, ES 2.10.1 I will let you know if I find something.

**Comment by nkajic** [comment]
Hi all, I have noticed that this problem only occurs when "Is Ancor" was not enabled on categories.  Is Anchor is hidden attribute on categories (in new version of Magento or Elastic Suite - not sure). 

I corrected by issue by changing is_anchor values in database on all categories that had is_anchor values.

1. ) Get your attribute id.
![Screenshot 2020-10-09 at 18 10 13](https://user-images.githubusercontent.com/3989412/95606500-b4b49500-0a5a-11eb-948b-a5c2d1ffc04a.png)

2.) Check values. I had both enabled and disabled.
![Screenshot 2020-10-09 at 18 11 16](https://user-images.githubusercontent.com/3989412/95606641-e6c5f700-0a5a-11eb-9d96-f2ce0bc2d7fd.png)

3.) Change all is_anchor values to enable
![Screenshot 2020-10-09 at 18 13 24](https://user-images.githubusercontent.com/3989412/95606828-34426400-0a5b-11eb-9f96-186f05e4160c.png)

4.) Verify after update
![Screenshot 2020-10-09 at 18 15 07](https://user-images.githubusercontent.com/3989412/95606994-6653c600-0a5b-11eb-941b-8fb5dd3de8ed.png)

5.) Reindex
6.) Flush chache
7.) Profit

**Comment by walkwizus** [comment]
Hi all,

In our case we have the same problem. Due to the call of a count on product collection before load Toolbar. In a custom theme when you display the total number of products (for example), Magento will load the product collection and the limit set in Toolbar block will not apply. Same thing on Luma theme in one column layout because the layers was loaded before the Toolbar.

The quick and dirty consists to clear the product collection at the final to enforce Magento to reload it and the toolbar will apply the limit.

Another approch will be move the pagination in _beforeLoad in another method because this method is called only once but in case of a count or another thing on this collection after this call, it's too late.

To clear the product collection just need to add a plugin.

Example:

app/code/Vendor/Namespace/etc/frontend/di.xml

`<?xml version="1.0"?>

<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:ObjectManager/etc/config.xsd">
    <type name="Magento\Catalog\Block\Product\ProductList\Toolbar">
        <plugin name="vendor_namespace_block_product_productlist_toolbar" type="Vendor\Namespace\Plugin\Block\Product\ProductList\Toolbar" />
    </type>
</config>`

app/code/Vendor/Namespace/Plugin/Block/Product/ProductList/Toolbar.php

`<?php

namespace Vendor\Namespace\Plugin\Block\Product\ProductList;

class Toolbar
{
    public function aroundSetCollection(\Magento\Catalog\Block\Product\ProductList\Toolbar $subject, callable $proceed, $collection)
    {
        $collection->clear();
        return $proceed($collection);
    }
}`

**Comment by refueledinc** [comment]
I too was having this issue on two categories and `is_anchor = 0` was the cause. Manually editing the database and setting to `1` as outlined by @nkajic fixed the issue for me.

I'm not sure if it's related, but the two categories in question also had erroneous rewrites that also needed to be fixed.

**Comment by groomershop-mt** [comment]
We also have still the problem at version 2.10.3 - after upgrade to that version on category pages the pagination is working correctly, but it still doesn't work on search result page.

**Comment by groomershop-mt** [comment]
Our problem was caused by not applied this commit to our theme: https://github.com/magento/magento2/commit/7035daca3f6b47038d120813ec7164b76ac4bc94#diff-7f64b8e3bdbb6406a30a1e53552a0f31dd086e65eaa6455aef7b81b6cf2389fd

**Comment by romainruaud** [comment]
Okay so after reading all of the answers, it seems to be due to categories having `is_anchor=0`, which should not be used with Elasticsuite (we arbitrary set it to 1 and we hide it).

Is that all ? If yes, may we close ?

Regards

**Comment by OvalMedia** [comment]
Negative. I have tried both options and it made not difference. 
The only real difference was the existence of the category filter on the page.
You should look into https://github.com/magento/magento2/pull/27165 as they have changed something significant there.
Also: https://github.com/Smile-SA/elasticsuite/issues/1925#issuecomment-682376815

**Comment by romainruaud** [comment]
Okay,

@vahonc can you try to handle this case where the category filter is hidden by the configuration ?

I think when it's "hidden", the filter model is not applied at all : 

https://github.com/Smile-SA/elasticsuite/blob/2.10.x/src/module-elasticsuite-catalog/Model/Layer/Filter/Category.php#L109

What we should obtain is : 

- the filter model is "still here", and he applies effectively filtering for the current category being viewed.
- but it's not displaying the childrens.

So maybe, we could : 

- enforce to have the filter model being there everytime, whatever is set in the configuration
- but we dismantle the display : maybe just returning an empty array in _getItems() is enough, this kind of things.

**Comment by romainruaud** [comment]
What we should do for this one.

**Create a new Filter model**

This filter model should be in the ElasticSuiteCatalog module, along with the others : https://github.com/Smile-SA/elasticsuite/tree/2.10.x/src/module-elasticsuite-catalog/Model/Product/Search/Request/Container/Filter

Reminder, it's injected like this : https://github.com/Smile-SA/elasticsuite/blob/2.10.x/src/module-elasticsuite-catalog/etc/elasticsuite_search_request.xml#L45

I think it should be injected only for the **catalog_category_view** query.

**Content of the Filter model**

He should be named `CurrentCategory` and implement the `getFilterQuery` method like this : 

- first, check if the category filter is disabled 
- if it's disabled, do the following : 
- retrieve the current category, like it's done [here](https://github.com/Smile-SA/elasticsuite/blob/2.10.x/src/module-elasticsuite-virtual-category/Model/Product/Search/Request/Container/Filter/CategoryBlacklist.php#L68) with `searchContext->getCurrentCategory()`
- use [this object](https://github.com/Smile-SA/elasticsuite/blob/25d03cc7800599a7dfb335c7c83ee82cb0848584/src/module-elasticsuite-catalog/Model/Category/Filter/Provider.php#L28) to build the query by calling `->getQueryFilter($category)`
- and finally return this query as the return of the implementation of `getFilterQuery` method.

**Comment by KZNcode** [comment]
@vahonc can you elaborate on what this does?

Also, do we need to take further steps (as in Admin) after applying this to the latest release version?

**Comment by vahonc** [comment]
Hello @KZNcode,

My fix added the ability to add/force the Current Category filter to the filters model to be being there every time when you are on a page of some category, regardless of what is set in the configuration (Stores > Config > Catalog > Catalog > Layered Navigation > Display Category filter). It fixed the issue of displaying all products from the catalog in the current category when the Display Category filter option set to No. 

For Magento 2.4.1 version there is no need for additional configuration in the Magento BO. As for Magento 2.4.2, I will try to test it as soon as possible.

BR,
Vadym

**Comment by KZNcode** [comment]
@vahonc Thanks for that.

**Comment by qsolutions-pl** [comment]
Had the same issue, it turned out that there was a bug in Magento 2 import proces which resulted in invalid "path" values in `catalog_category_entity` table.
Once we fixed this on our end, there are no more issues as described.

**Comment by ProxiBlue** [comment]
2025, and this bug still exists?


```
public function getFilterQuery()
    {
        $query = null;

        //if (false === $this->isDisplayCategoryFilter() && $this->searchContext->getCurrentCategory()) {
            $query = $this->filterProvider->getQueryFilter($this->searchContext->getCurrentCategory());
        //}

        return $query;
    }
```

**Comment by rbayet** [comment]
@ProxiBlue 2025 and **developers** don't know that bugs can be fixed (see the mention of the merged PR) and then creep back up again ?
2025 and commenting on a closed issue of a free and open-source project with the aggressiveness of a 19th century landlord coming to collect the past due rents is still a thing ?

So, we can keep going on with the passive-agressiveness (I'm a black belt in it) or we can start being constructive (like @OvalMedia is by providing at least a bit of context).
I'll open a new issue linked to this one and we'll try to reproduce the problem.

Regards,

**Comment by rbayet** [comment]
@OvalMedia @ProxiBlue please provide any additional context in the new issue #3506, particularly
* if this seen in developer or production or both
* the magento type (enterprise or community)
* a list of third party modules, particularly those related to layered navigation or category related features that are installed on the problematic environments
